### PR TITLE
[WIP] [Coconut] Data Storage features

### PR DESCRIPTION
**Pull request by suresh-lc**
# Description

Base code for "Data Storage" feature.

Fixes # (issue)
N/A 

## Type of change
Feature addition

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

**Test Configuration**:
* Firmware version: Ubuntu 18.04
* Hardware: x86-64
* Edge Orchestration Release: Data Storage

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: GoMain/src/main/main.go [hunk 1]
```diff
@@ -75,6 +85,9 @@ func main() {
 		log.Fatalf("[%s] Orchestaration initalize fail : %s", logPrefix, err.Error())
 	}
 	sigmgr.Watch()
+	for {
+		time.Sleep(1000)
```
Comment (MoonkiHong) [inline GoMain/src/main/main.go:89]: Please see the merged PR https://github.com/lf-edge/edge-home-orchestration-go/pull/108/files. It is now deprecated. (Recommend you to sync up with the existing codes in the main-stream branch).

File: doc/datastorage_setup [hunk 2]
```diff
@@ -0,0 +1,36 @@
+Installing ZeroMQ:
+
+# run in sudo
+# Before installing, make sure you have installed all the needed packages
+1.sudo apt-get install libtool pkg-config build-essential autoconf automake
```
Comment (MoonkiHong) [inline doc/datastorage_setup:5]: Revise the instruction applying the coding style font like `sudo apt-get...`.

File: doc/datastorage_setup [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
```
Comment (MoonkiHong) [inline doc/datastorage_setup:6]: Is this an intentional blank line? If not, please remove it.

File: doc/datastorage_setup [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+1. Download the tar from https://github.com/zeromq/libzmq/tags/v4.2.2
+2. tar xvf libzmq-4.2.2.tar.gz
+3. cd libzmq-4.2.2/
+4. ./autogen.sh
+5. sudo apt-get update
+6. sudo apt-get install -y libtool pkg-config build-essential autoconf automake uuid-dev
+7. ./configure
+8. sudo make install
+9. sudo ldconfig
+# Check installed
```
Comment (MoonkiHong) [inline doc/datastorage_setup:29]: Better to put an blank line between Line 28 and 29. What do you think?

File: doc/datastorage_setup.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
+# ZeroMQ installation on Linux
+
+## Quick start ##
+This section provides how to download and install zeromq
+
+#### 1. Install the dependency packages
+- sudo apt-get install libtool pkg-config build-essential autoconf   automake
+
```
Comment (MoonkiHong) [inline doc/datastorage_setup.md:8]: Better to remove this blank-line if this is not your intention. What do you think?

File: doc/datastorage_setup.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
+# ZeroMQ installation on Linux
+
+## Quick start ##
+This section provides how to download and install zeromq
+
+#### 1. Install the dependency packages
+- sudo apt-get install libtool pkg-config build-essential autoconf   automake
```
Comment (MoonkiHong) [inline doc/datastorage_setup.md:7]: Better to apply the coding style font like `sudo apt-get...`.

File: doc/datastorage_setup.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+4. ./autogen.sh 
+If it succeeds, you can see ./configure file generated
+5. ./configure
+6. make && make check
+7. sudo make install
+8. sudo ldconfig
+
+#### 3. Install zeromq
+After installing the Dependency packages, zero mq can be isntalled with the following steps
+
```
Comment (MoonkiHong) [inline doc/datastorage_setup.md:24]: Any intention to put a blank line here?

**Review by MoonkiHong — CHANGES_REQUESTED** [review]
@suresh-lc Please see my comments. Plus, could you put the git log history that reflects all the developments during the incubation stage?

**Comment by MoonkiHong** [comment]
@suresh-lc Plus, please add the description of the `How Has This Been Tested?` and submit the regarding Issue in GitHub and link it to this PR.

File: GoMain/src/main/main.go [hunk 8]
```diff
@@ -60,9 +66,13 @@ const (
 	certificateFilePath    = edgeDir + "/data/cert"
 	containerWhiteListPath = edgeDir + "/data/cwl"
 	passPhraseJWTPath      = edgeDir + "/data/jwt"
+	binPath                = "GoMain/bin"
```
Comment (t25kim) [inline GoMain/src/main/main.go:69]: What do you think of using edgeDir as below?
```suggestion
	binPath                = edgeDir + "/bin"
```
If this is more reasonable, the YamlFileName and the ConfigFileName are should be changed as well.

File: GoMain/src/main/main.go [hunk 9]
```diff
@@ -60,9 +66,13 @@ const (
...
+	YamlFileName           = "/GoMain/bin/res/sample-json-device.yaml"
+	ConfigFileName         = "/GoMain/bin/res/configuration.toml"
 
 	cipherKeyFilePath = edgeDir + "/user/orchestration_userID.txt"
 	deviceIDFilePath  = edgeDir + "/device/orchestration_deviceID.txt"
+	serviceName       = "datastorage"
```
Comment (t25kim) [inline GoMain/src/main/main.go:75]: I'm a little bit worried that this main file seems to be only for the datastorage service.

**Review by t25kim — COMMENTED** [review]
The container that are the result of building the edge-orchestration doesn't run correctly.
You can find the log below.
```
2020/10/05 09:47:03 route.go:78: {APIV1ScoringmgrScoreLibnameGet GET /api/v1/scoringmgr/score 0x9d74a0}
2020/10/05 09:47:03 route.go:78: {APIV1ScoringmgrResourceGet GET /api/v1/scoringmgr/resource 0x9d7500}
2020/10/05 09:47:03 route.go:78: {APIV1RequestServicePost POST /api/v1/orchestration/services 0x9d47a0}
2020/10/05 09:47:03 route.go:78: {APIV1RequestSecuremgrPost POST /api/v1/orchestration/securemgr 0x9d4800}
2020/10/05 09:47:03 route.go:109: ListenAndServe_For_Inter
2020/10/05 09:47:03 route.go:113: ListenAndServe
2020/10/05 09:47:03 main.go:175: interface orchestration init done
2020/10/05 09:47:03 uuidmapping.go:83: [storagemgr] MapYamlFile() +
2020/10/05 09:47:03 uuidmapping.go:50: [storagemgr] GetUuid() +
2020/10/05 09:47:03 uuidmapping.go:57: [storagemgr] The device id obtained : edge-orchestration-3125da9e-1e9a-41aa-ac83-004725eb2d1e
2020/10/05 09:47:03 uuidmapping.go:75: [storagemgr] Final UUid obtained is 3125da9e-1e9a-41aa-ac83-004725eb2d1e-
2020/10/05 09:47:03 uuidmapping.go:88: [storagemgr] the path before directory coming out is /edge-orchestration
2020/10/05 09:47:03 uuidmapping.go:97: [storagemgr] the path is  /
2020/10/05 09:47:03 uuidmapping.go:100: open //GoMain/bin/res/sample-json-device.yaml: no such file or directory
```

File: GoMain/src/main/main.go (hunk 9 above)
Comment (suresh-lc) [reply GoMain/src/main/main.go:75]: Incorporated

File: GoMain/src/main/main.go (hunk 8 above)
Comment (suresh-lc) [reply GoMain/src/main/main.go:69]: Ok will consider for next PR

File: GoMain/src/main/main.go [hunk 10]
```diff
@@ -160,6 +173,25 @@ func orchestrationInit() error {
 	restEdgeRouter.Start()
 
 	log.Println(logPrefix, "orchestration init done")
+	storagemgr.MapYamlFile(YamlFileName)
```
[the remark this answers is not in the document]
Comment (suresh-lc) [reply GoMain/src/main/main.go:173]: We already have thought on the same line. We will take this up as next priority item once base code is pushed

File: doc/datastorage_setup.md [hunk 11]
```diff
@@ -0,0 +1,53 @@
+# ZeroMQ installation on Linux
```
[the remark this answers is not in the document]
Comment (suresh-lc) [reply doc/datastorage_setup.md:1]: Changes done

File: doc/datastorage_setup [hunk 12]
```diff
@@ -0,0 +1,36 @@
+Installing ZeroMQ:
```
[the remark this answers is not in the document]
Comment (suresh-lc) [reply doc/datastorage_setup:1]: Done

File: build.sh [hunk 13]
```diff
@@ -33,6 +33,8 @@ PKG_LIST=(
         "controller/servicemgr/executor/containerexecutor"
         "controller/servicemgr/executor/nativeexecutor"
         "controller/servicemgr/notification"
+        "controller/storagemgr/util"
+  		"controller/storagemgr/storagedriver"
```
[the remark this answers is not in the document]
Comment (suresh-lc) [reply build.sh:37]: Done

File: GoMain/src/main/main.go [hunk 14]
```diff
@@ -160,6 +173,25 @@ func orchestrationInit() error {
...
+	if err != nil {
+		log.Fatalf("[%s] Error in directory change: %s", logPrefix, err.Error())
+	}
+	log.Println(logPrefix, "Path set is now", pwd)
+	sd := storagedriver.StorageDriver{}
+	startup.Bootstrap(serviceName, device.Version, &sd)
+
+	// TODO remove this line
+	// this line is for wait to initialize the mDNS server.
+	time.Sleep(time.Second * 2)
```
[the remark this answers is not in the document]
Comment (suresh-lc) [reply GoMain/src/main/main.go:191]: Fixed

**Comment by suresh-lc** [comment]
Changes done. Request review.

File: GoMain/src/main/main.go (hunk 1 above)
Comment (suresh-lc) [reply GoMain/src/main/main.go]: Done

File: doc/datastorage_setup (hunk 2 above)
Comment (suresh-lc) [reply doc/datastorage_setup]: Done

File: doc/datastorage_setup (hunk 3 above)
Comment (suresh-lc) [reply doc/datastorage_setup]: Fixed

File: doc/datastorage_setup (hunk 4 above)
Comment (suresh-lc) [reply doc/datastorage_setup]: Done

File: doc/datastorage_setup.md (hunk 5 above)
Comment (suresh-lc) [reply doc/datastorage_setup.md]: Done

File: doc/datastorage_setup.md (hunk 6 above)
Comment (suresh-lc) [reply doc/datastorage_setup.md]: Done

File: doc/datastorage_setup.md (hunk 7 above)
Comment (suresh-lc) [reply doc/datastorage_setup.md]: Done

File: GoMain/bin/res/configuration.toml [hunk 15]
```diff
@@ -0,0 +1,78 @@
+[Writable]
```
Comment (t25kim) [inline GoMain/bin/res/configuration.toml:1]: Is this a sample for the datastorage configuration?
Why don't you put the sample files in the [samples/datastorage folder](https://github.com/lf-edge/edge-home-orchestration-go/tree/master/samples)?

File: GoMain/src/main/main.go [hunk 16]
```diff
@@ -21,6 +21,8 @@ package main
 import (
 	"errors"
 	"log"
+	"os"
+	"time"
```
Comment (t25kim) [inline GoMain/src/main/main.go:25]: Please check the build test and upload the PR.
```
----------------------------------------
 Create Executable binary from GoMain
----------------------------------------
go build -a -ldflags '-extldflags "-static" -X main.version= -X main.commitID=6621c98 -X main.buildTime=20201006.0826 -X main.buildTags=' -o /home/t25kim/work/edge-home-orchestration-go/GoMain/bin/edge-orchestration main || exit 1
# main
GoMain/src/main/main.go:25:2: imported and not used: "time"
Makefile:65: recipe for target 'build-binary' failed
```

**Review by t25kim — COMMENTED** [review]
I can't test the datastorage function since the problem I mentioned above is not solved yet.

File: doc/datastorage_setup.md [hunk 17]
```diff
@@ -4,50 +4,72 @@
 This section provides how to download and install zeromq
 
 #### 1. Install the dependency packages
-- sudo apt-get install libtool pkg-config build-essential autoconf   automake
-
-- sudo apt-get install libzmq3-dev
+```
+sudo apt-get install libtool pkg-config build-essential autoconf   automake
```
Comment (MoonkiHong) [inline doc/datastorage_setup.md]:
```suggestion
sudo apt-get install libtool pkg-config build-essential autoconf automake
```
@suresh-lc What about this?

File: doc/datastorage_setup.md [hunk 18]
```diff
@@ -4,50 +4,72 @@
...
+sudo apt-get install libzmq3-dev
+```
 
 #### 2. Install libsodium (Zero MQ has dependcy on this package)
-1. Download the tar from https://libsodium.gitbook.io/doc/installation (Version: 1.0.18)
-2. tar -xvf tar -xvf libsodium_1.0.18.orig.tar.gz 
-3. cd libsodium-1.0.18
-4. ./autogen.sh 
+
```
Comment (MoonkiHong) [inline doc/datastorage_setup.md]:
```suggestion
```
@suresh-lc If there is no intention to add a blank-line, then what about removing it like this?

**Review by MoonkiHong — CHANGES_REQUESTED** [review]
Please see my additional comments.

File: GoMain/bin/res/configuration.toml (hunk 15 above)
Comment (MoonkiHong) [reply GoMain/bin/res/configuration.toml:1]: +1 as @t25kim pointed out.

File: GoMain/bin/res/docker/configuration.toml [hunk 19]
```diff
@@ -0,0 +1,70 @@
+[Writable]
```
Comment (MoonkiHong) [inline GoMain/bin/res/docker/configuration.toml:1]: @suresh-lc This is also the same as @t25kim pointed out for `GoMain/bin/res/configuration.toml` that this is a sample so it is better to place this into [the samples/datastorage folder](https://github.com/lf-edge/edge-home-orchestration-go/tree/master/samples). What do you think?

File: GoMain/bin/res/sample-json-device.yaml [hunk 20]
```diff
@@ -0,0 +1,61 @@
+name: "7e841a12-a083-48fe-8040-83815e0b7410"
```
Comment (MoonkiHong) [inline GoMain/bin/res/sample-json-device.yaml:1]: @suresh-lc Same opinion as `GoMain/bin/res/configuration.toml` and `GoMain/bin/res/docker/configuration.toml`.

**Review by MoonkiHong — CHANGES_REQUESTED** [review]
Please see my comments.

File: GoMain/src/main/main.go (hunk 8 above)
Comment (MoonkiHong) [reply GoMain/src/main/main.go:69]: Created the new issue #136. ^^

File: GoMain/src/main/main.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (MoonkiHong) [reply GoMain/src/main/main.go:174]: Created a new issue #137. ^^

**Review by MoonkiHong — COMMENTED** [review]
Verified.

File: GoMain/bin/res/configuration.toml (hunk 15 above)
Comment (suresh-lc) [reply GoMain/bin/res/configuration.toml:1]: Good point, incorporated

File: doc/datastorage_setup.md (hunk 17 above)
Comment (suresh-lc) [reply doc/datastorage_setup.md]: Incorporated

File: samples/datastorage/sample-json-device.yaml [hunk 21]
```diff
@@ -0,0 +1,61 @@
+name: "7e841a12-a083-48fe-8040-83815e0b7410"
+manufacturer: "Intel Corp."
```
Comment (MoonkiHong) [inline samples/datastorage/sample-json-device.yaml:2]: @suresh-lc If this is a sample, what about anonymizing the manufacturer name for the device? Like _Vendor A_

**Review by MoonkiHong — COMMENTED** [review]
Please see my additional 1 comment.

**Comment by MoonkiHong** [comment]
@suresh-lc Please check out the DCO verification.

**Review by MoonkiHong — COMMENTED** [review]
Verified the revised code.

File: samples/datastorage/sample-json-device.yaml (hunk 21 above)
Comment (suresh-lc) [reply samples/datastorage/sample-json-device.yaml:2]: OK sure

**Comment by suresh-lc** [comment]
Have done necessary and pushed.

**Comment by MoonkiHong** [comment]
Build stuck (reported by @t25kim )
```
2020/10/07 11:20:00 uuidmapping.go:110: [storagemgr] MapConfigFile() +
2020/10/07 11:20:00 uuidmapping.go:116: [storagemgr] the path is  /var/edge-orchestration/datastorage/configuration.toml
2020/10/07 11:20:00 uuidmapping.go:144: [storagemgr] Returning from the uuidmapping.go
Init: useRegistry:  profile:  confDir:
Loading configuration from: /edge-orchestration/res/configuration.toml

Load configuration from local file and bypassing registration in Registry...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x9bca1d]

goroutine 1 [running]:
github.com/edgexfoundry/device-sdk-go.NewService(0xb35747, 0xb, 0xb3face, 0x1a, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/t25kim/work/edge-home-orchestration-go/vendor/src/github.com/edgexfoundry/device-sdk-go/service.go:287 +0x23d
github.com/edgexfoundry/device-sdk-go/pkg/startup.startService(0xb35747, 0xb, 0xb3face, 0x1a, 0xc0d880, 0xc000096900, 0x0, 0x40fb58)
        /home/t25kim/work/edge-home-orchestration-go/vendor/src/github.com/edgexfoundry/device-sdk-go/pkg/startup/bootstrap.go:46 +0xd4
github.com/edgexfoundry/device-sdk-go/pkg/startup.Bootstrap(0xb35747, 0xb, 0xb3face, 0x1a, 0xc0d880, 0xc000096900)
        /home/t25kim/work/edge-home-orchestration-go/vendor/src/github.com/edgexfoundry/device-sdk-go/pkg/startup/bootstrap.go:39 +0x10a
main.orchestrationInit(0xc000000180, 0xc00025ff78)
        /home/t25kim/work/edge-home-orchestration-go/GoMain/src/main/main.go:175 +0x89c
main.main()
        /home/t25kim/work/edge-home-orchestration-go/GoMain/src/main/main.go:84 +0x26

```

**Comment by MoonkiHong** [comment]
@suresh-lc Just reminder. Please refer to the suggested code refactoring from @t25-kim and let us know the update for this PR. Plus, we would like to close this PR for doing that, and squash the current commit w.r.t. the contributed sub-features for Data Storage according to `https://github.com/lf-edge/edge-home-orchestration-go/blob/master/.github/CONTRIBUTING.md` Section 11.

**Comment by MoonkiHong** [comment]
Will re-submit the following PR with the proper code refactor suggested by @t25kim. So I close this PR.

**Pull request closed without merging**